fix: support multiple versions of Elasticsearch logger

Revolyssup opened the pull request:
Fixes 
Error comes when using elasticsearch version 9+
![Screenshot From 2025-06-23 12-29-02](https://github.com/user-attachments/assets/00fc81be-2134-496c-9dce-c8196dcf87be)

We want to remove the compat header and have a general support for all major versions.

For elasticsearch logger plugin comptability check. Following are the observations and conclusion based on reading docs and self testing.

The `type` field issue:
Before version 7, the `type` field was required in json schema https://github.com/elastic/elasticsearch/blob/8453f7701de25c467ce08088ddddac185cb8028c/rest-api-spec/src/main/resources/rest-api-spec/api/create.json#L21.
The current tests without `type` fail on version 6(here it's 6.7) with the following error

Batch Processor[elasticsearch-logger] failed to process entries: elasticsearch server returned status: 400, body: {"error":{"root_cause":[{"type":"action_request_validation_exception","reason":"Validation Failed: 1: type is missing;"}]

Adding a _type in the code fixes the above test.
Conclusion
As suggested in the migration guide, we can hardcode the `type` to "_doc" in versions prior to 7 and on 7+ we can omit `type` completely
Indices created in 6.x only allow a single-type per index. Any name can be used for the type, but there can be only one. The preferred type name is _doc, so that index APIs have the same path as they will have in 7.0: PUT {index}/_doc/{id} and POST {index}/_doc


The `id` field issue:
Another problem was detected while running tests on older version(6), when _id was not passed along with _index,. the following error was reported:

Batch Processor[elasticsearch-logger] failed to process entries: elasticsearch server returned status: 400, body: {"error":{"root_cause":[{"type":"action_request_validation_exception","reason":"Validation Failed: 1: an id must be provided if version type or value are set;"}],"

In older version, _id is also required https://github.com/elastic/elasticsearch/blob/8453f7701de25c467ce08088ddddac185cb8028c/rest-api-spec/src/main/resources/rest-api-spec/api/create.json#L9
But I when I used index API instead of create API, request passed even without _id for all versions 6,7,8

-            create = {
+            index = {
                 _index = conf.field.index,
-                _type = conf.field.type
+                _type = "_doc",
             }
         }) .. "\n" ..


The doc says
A create action fails if a document with the same ID already exists in the target An index action adds or replaces a document as necessary.
An answer in the discuss forum says
Neither is used to create an index (although they may do so as a side-effect of inserting the first document into a new index). "index" indexes a document. "create" only indexes a document if it doesn't already exist. It used to be that "create" was faster if you knew that your docs didn't already exist. But now, since versioning was added, "index" performs as well as "create"
Another answer suggests
you use create if a doc with that id already exists, if you use index it will create a document or update it if xisting.
Conclusion: To me it looks okay to replace create with index



Conclusion:
We can remove `type` from plugin configuration and for detected version below 7, hardcode it as `_doc`
We can replace Create with Index in the request body.
We run all tests on version 8 and my PR adds a comptability test with version 9. We can add two more compatibilty tests with 6 and 7 to confirm our changes.

Result: We will be able to support ES versions 6,7,8,9

References and other relevant information:
https://www.elastic.co/guide/en/elasticsearch/reference/6.7/release-notes-6.7.0.html. Search "Deprecate types" in this doc.
From reading this issue: https://github.com/elastic/elasticsearch/issues/35190
Though a temporary parameter was added `include_type_name` which defaults to true between 6.7 upto 7 meaning by default `type` was being sent but it could be turned off.
Then between 7 and 8 this field was default to false meaning by default the `type` will no longer be sent and a deprecated warning will come.
Then after 8+ this include_type_name was removed and `type` was no longer supported.
References:
https://www.elastic.co/docs/reference/elasticsearch/rest-apis/compatibility
https://www.elastic.co/guide/en/elasticsearch/reference/8.18/rest-api-compatibility.html
https://github.com/elastic/elasticsearch-ruby/issues/2665
### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

<!--

Note

1. Mark the PR as draft until it's ready to be reviewed.
2. Always add/update tests for any changes unless you have a good reason.
3. Always update the documentation to reflect the changes made in the PR.
4. Make a new commit to resolve conversations instead of `push -f`.
5. To resolve merge conflicts, merge master instead of rebasing.
6. Use "request review" to notify the reviewer after making changes.
7. Only a reviewer can mark a conversation as resolved.

-->

nic-6443 commented on ci/pod/docker-compose.plugin.yml:
```diff
@@ -225,6 +225,66 @@ services:
       http.port: 9201
       xpack.security.enabled: 'true'
 
+  elasticsearch-noauth-2:
```
Running so many Elasticsearch containers simultaneously in CI may strain computing resources, and causing tests to fail.

Revolyssup replied:
i removed two of these containers(not needed)

nic-6443 commented on apisix/plugins/elasticsearch-logger.lua:
```diff
@@ -124,32 +124,94 @@ function _M.check_schema(conf, schema_type)
         return false, str_format("create http error: %s", err)
     end
-
+    set_version(conf)
```
we already call `set_version` here, why still need it in `access` phase?

Revolyssup replied:
this function is run later in batch processor. But we need the version in the log phase when entry is created. That is why we need it in access phase.

Revolyssup commented on ci/pod/docker-compose.plugin.yml:
```diff
@@ -225,6 +225,66 @@ services:
+      xpack.security.enabled: 'true'
+
+  elasticsearch-auth-4:
+    image: docker.elastic.co/elasticsearch/elasticsearch:6.7.0
```
i removed two of these containers(not needed)

Revolyssup commented on apisix/plugins/elasticsearch-logger.lua:
```diff
@@ -124,32 +124,94 @@ function _M.check_schema(conf, schema_type)
         core.json.encode(entry) .. "\n"
 end
 
+local function set_version(conf)
```
done

Revolyssup commented on apisix/plugins/elasticsearch-logger.lua:
```diff
@@ -124,32 +124,94 @@ function _M.check_schema(conf, schema_type)
+        end
+        local major_version, err = get_es_major_version(selected_endpoint_addr, conf)
+        if err then
+            return false, str_format("failed to get Elasticsearch version: %s", err)
```
done

Revolyssup commented:
> > > Do we need to remove `type` field in this PR?
> > 
> > 
> > Yes. If we don't remove `type` field then we will need the compat headers that I removed
> 
> pls update the doc too

Removed `type` from docs

membphis reviewed (approved):
LGTM
